File: runner-scripts/dubbing-engine/python/separate.py
#!/usr/bin/env python3
"""
Dubbing Engine – Stage 2: Vocal Separation (Demucs)

Usage:
    python separate.py --input audio_stereo.wav --output /path/to/output_dir

Produces:
    output_dir/vocals.wav
    output_dir/no_vocals.wav
"""
import argparse
import json
import os
import sys
import time


def main():
    parser = argparse.ArgumentParser(description="Demucs vocal separation")
    parser.add_argument("--input", required=True, help="Input stereo WAV file")
    parser.add_argument("--output", required=True, help="Output directory")
    args = parser.parse_args()

    input_file = args.input
    output_dir = args.output

    if not os.path.exists(input_file):
        print(f"[SEPARATE] Error: Input not found: {input_file}", file=sys.stderr)
        sys.exit(1)

    os.makedirs(output_dir, exist_ok=True)

    print(f"[SEPARATE] Separating vocals from: {input_file}")
    sys.stdout.flush()

    try:
        import torch
        import demucs.separate

        # Demucs API call
        demucs.separate.main([
            "--two-stems", "vocals",
            "-o", output_dir,
            input_file,
        ])

        # Find the output (Demucs creates subdir with model name)
        model_dirs = [d for d in os.listdir(output_dir)
                      if os.path.isdir(os.path.join(output_dir, d))]
        model_dir = model_dirs[0] if model_dirs else "htdemucs"
        src_dir = os.path.join(output_dir, model_dir)

        # Copy/rename to standard locations
        vocals_src = os.path.join(src_dir, "vocals.wav")
        no_vocals_src = os.path.join(src_dir, "no_vocals.wav")

        if os.path.exists(vocals_src):
            import shutil
            shutil.copy2(vocals_src, os.path.join(output_dir, "vocals.wav"))
            print(f"[SEPARATE] Vocals saved: {os.path.join(output_dir, 'vocals.wav')}")

        if os.path.exists(no_vocals_src):
            import shutil
            shutil.copy2(no_vocals_src, os.path.join(output_dir, "no_vocals.wav"))
            print(f"[SEPARATE] Background saved: {os.path.join(output_dir, 'no_vocals.wav')}")

    except ImportError as e:
        print(f"[SEPARATE] Dependency error: {e}", file=sys.stderr)
        # Fallback: just copy input as vocals, create silent background
        import shutil
        shutil.copy2(input_file, os.path.join(output_dir, "vocals.wav"))
        # Create a silent placeholder for no_vocals
        duration = 1.0
        try:
            import subprocess
            result = subprocess.run(
                ["ffprobe", "-v", "error", "-show_entries", "format=duration",
                 "-of", "default=noprint_wrappers=1:nokey=1", input_file],
                capture_output=True, text=True, timeout=15
            )
            if result.stdout.strip():
                duration = float(result.stdout.strip())
        except Exception:
            pass

        silent_file = os.path.join(output_dir, "no_vocals.wav")
        try:
            import subprocess
            subprocess.run(
                ["ffmpeg", "-y", "-f", "lavfi", "-t", str(duration),
                 "-i", "anullsrc=r=44100:cl=stereo",
                 "-acodec", "pcm_s16le", silent_file],
                capture_output=True, timeout=30
            )
        except Exception:
            # Can't create silent audio - that's ok, mix stage handles it
            pass

    except Exception as e:
        print(f"[SEPARATE] Error: {e}", file=sys.stderr)
        sys.exit(1)

    # Verify output
    vocals_file = os.path.join(output_dir, "vocals.wav")
    bg_file = os.path.join(output_dir, "no_vocals.wav")
    result = {
        "vocals_exist": os.path.exists(vocals_file),
        "no_vocals_exist": os.path.exists(bg_file),
    }
    print(f"[SEPARATE] Result: {json.dumps(result)}")


if __name__ == "__main__":
    main()

File: runner-scripts/dubbing-engine/lib/separate.js
/**
 * Stage 2 – Separate (Demucs vocal separation)
 *
 * Calls a Python script that uses Demucs to split:
 *   - vocals.wav
 *   - no_vocals.wav (drums + bass + other = background)
 */
const path = require('path');
const fs = require('fs');
const utils = require('./utils');

function getPythonScriptPath() {
  return path.resolve(__dirname, '..', 'python', 'separate.py');
}

async function separate(workDir, manifest) {
  const inputStereo = path.join(workDir, 'audio_stereo.wav');
  const outputDir   = path.join(workDir, 'separated');

  utils.logStep('SEPARATE', `Input: ${inputStereo}`);

  if (!fs.existsSync(inputStereo)) {
    console.log('[SEPARATE] No stereo audio found – skipping vocal separation');
    return { skipped: true, reason: 'No audio track' };
  }

  // Check if Demucs is available by trying import
  let demucsAvailable = false;
  try {
    await utils.runProcess(utils.getPython(), [
      '-c', 'import demucs; print(demucs.__version__)'
    ], { stdio: 'pipe', timeoutMs: 10000, logLabel: 'SEPARATE' });
    demucsAvailable = true;
    console.log('[SEPARATE] Demucs detected – running separation…');
  } catch {
    console.log('[SEPARATE] Demucs not installed – will copy audio as-is');
  }

  utils.ensureDir(outputDir);

  if (demucsAvailable) {
    // Run Demucs
    const pythonScript = getPythonScriptPath();
    await utils.runPython(pythonScript, [
      '--input', inputStereo,
      '--output', outputDir,
    ], { logLabel: 'SEPARATE', timeoutMs: 3600000 });

    // Verify outputs – Demucs may use subdir like "htdemucs"
    const vocals    = path.join(outputDir, 'vocals.wav');
    const noVocals  = path.join(outputDir, 'no_vocals.wav');

    if (!fs.existsSync(vocals)) {
      const subDirs = fs.readdirSync(outputDir).filter(d => fs.statSync(path.join(outputDir, d)).isDirectory());
      for (const sub of subDirs) {
        const subVocals = path.join(outputDir, sub, 'vocals.wav');
        if (fs.existsSync(subVocals)) {
          utils.copyFile(subVocals, vocals);
        }
        const subNoVocals = path.join(outputDir, sub, 'no_vocals.wav');
        if (fs.existsSync(subNoVocals)) {
          utils.copyFile(subNoVocals, noVocals);
        }
        const subDrums = path.join(outputDir, sub, 'drums.wav');
        const subBass = path.join(outputDir, sub, 'bass.wav');
        const subOther = path.join(outputDir, sub, 'other.wav');
        if (!fs.existsSync(subNoVocals) && fs.existsSync(subDrums) && fs.existsSync(subBass) && fs.existsSync(subOther)) {
          const { execSync } = require('child_process');
          execSync(
            `${utils.FFMPEG} -y -i ${utils.quote(subDrums)} -i ${utils.quote(subBass)} -i ${utils.quote(subOther)} ` +
            `-filter_complex "[0:a][1:a][2:a]amix=inputs=3:duration=longest" -ac 2 ${utils.quote(noVocals)}`,
            { stdio: 'inherit', timeout: 300000 }
          );
        }
      }
    }
  } else {
    // Demucs not available – copy stereo as "vocals", create silent "no_vocals"
    utils.copyFile(inputStereo, path.join(outputDir, 'vocals.wav'));

    const duration = utils.getVideoDuration(inputStereo) || 0;
    try {
      const { execSync } = require('child_process');
      execSync(
        `${utils.FFMPEG} -y -f lavfi -t ${duration} -i anullsrc=r=44100:cl=stereo -acodec pcm_s16le ${utils.quote(path.join(outputDir, 'no_vocals.wav'))}`,
        { stdio: 'inherit', timeout: 120000 }
      );
    } catch {
      console.log('[SEPARATE] Could not create silent background audio');
    }
  }

  const result = {
    vocals_exist: fs.existsSync(path.join(outputDir, 'vocals.wav')),
    no_vocals_exist: fs.existsSync(path.join(outputDir, 'no_vocals.wav')),
    fallback: !demucsAvailable,
  };
  const flag = !demucsAvailable ? ' ⚠️ (fallback – Demucs not installed)' : '';
  console.log(`[SEPARATE] ✅ Done – Vocals: ${result.vocals_exist}, Background: ${result.no_vocals_exist}${flag}`);
  return result;
}

module.exports = { separate };
